### PR TITLE
[PWA-1947] Gift options on Order Level (Commerce only)

### PR DESCRIPTION
**Pull request by mikhaelbois**
## Description

Integration the Gift Message in Order Level feature in the Cart and in the Checkout.
Note that this is currently a Commerce Only feature as the [setGiftOptionsOnCart](https://devdocs.magento.com/guides/v2.4/graphql/mutations/set-gift-options.html) mutation is Commerce Only at the moment.

This is a follow-up of https://github.com/magento/pwa-studio/pull/3435/.

## Related Issue

Closes https://jira.corp.magento.com/browse/PWA-1947.

#### Test scenario(s) for any existing impacted features/areas

Preconditions
1. Adobe Commerce Store with PWA Storefront
2. Gift Messages on order Level is enabled in Stores > Settings > Configuration > Sales> Gift Options
3. Gift Messages for Order Items is disabled in Stores > Settings > Configuration > Sales> Gift Options
4. Gift Wrapping is disabled in Stores > Settings > Configuration > Sales> Gift Options
5. Gift Receipt is enabled
6. Printed Card is enabled

- Users can add gift messages to the oder level (not on order item level )
- User sees no option for gift receipt when Gift Receipt is disabled in Admin Config
- User sees no option for printed card when Printed Card is disabled in Admin Config
- User sees cart prices change when Printed Card has a price defined in Admin Config
- Codebase no longer contains mocked gift options data
- Cypress tests are running

## Checklist

-   I have added tests to cover my changes, if necessary.
-   I have added translations for new strings, if necessary.
-   I have updated the documentation accordingly, if necessary.

### CODE REVIEW

**Review by pankhurigoel — APPROVED** [review]
Looks good overall. Thanks.

**Review by pankhurigoel — CHANGES_REQUESTED** [review]
Looks good overall. Please fix the coverages on following files. `packages/venia-ui/lib/components/CartPage/PriceAdjustments/GiftOptions/giftOptions.js`, `packages/venia-ui/lib/components/CartPage/PriceAdjustments/giftOptionsSection.js`, `packages/venia-ui/lib/components/CartPage/PriceAdjustments/giftOptionsSection.ee.js`

**Comment by mikhaelbois** [comment]
@pankhurigoel Tests added. Coverage looks fine now.

**Comment by pankhurigoel** [comment]
> yarn run validate-queries

@mikhaelbois try running the `yarn run validate-queries` command locally and fix the issues.

**Comment by mikhaelbois** [comment]
@pankhurigoel Here you go. Sorry about that.

**Review by pankhurigoel — DISMISSED** [review]
Looks perfect!! Thanks.

**Comment by JustinConabreeAbsolunet** [comment]
⚠️ QA Smoke test passed. However looking at the database, there's multiple gift message entries created (some partial entries while the user is typing)
⚠️ Cypress tests pass with the exception of the contact page
`Expected to find element: [data-cy="ContactPage-formContainer"] input[name="name"], but never found it.`

**Comment by mikhaelbois** [comment]
@JustinConabreeAbsolunet 
I've updated the way the mutation works.
It will now save to cart only when we click on "Review Order" in the Checkout Page.
The other times, it will only save in the cache instead.
It should reduce the number of entries by a lot.

Also fixed the ContactPage Cypress test.

**Comment by JustinConabreeAbsolunet** [comment]
@mikhaelbois Getting closer!
✅ Cypress
⚠️ QA issue fixed but there's a new one. Null value saved in backend even when form is untouched. Could be a good idea to keep a "modified" state for the form an only send null if a non-null value was previously saved

**Comment by mikhaelbois** [comment]
@JustinConabreeAbsolunet 
Added a comparaison on previous form data and current form data.
If the data is not the same, the form will be submitted.
This way, when you start fresh and don't use the form, it will not be submitted.

**Comment by JustinConabreeAbsolunet** [comment]
@mikhaelbois Progress! Now working correctly in Checkout. However, if you add the gift message details in Cart and then go to Checkout, it doesn't get saved when placing the order

File: packages/peregrine/lib/talons/CartPage/PriceAdjustments/GiftOptions/useGiftOptions.js
Comment (pedrochiossi) [inline packages/peregrine/lib/talons/CartPage/PriceAdjustments/GiftOptions/useGiftOptions.js:85]: Can't we just use cart.gift_message as dependencies here?

Comment (mikhaelbois) [reply packages/peregrine/lib/talons/CartPage/PriceAdjustments/GiftOptions/useGiftOptions.js:85]: That's what the linter gave me.
From my understanding, if we only use `cart.gift_message`, it won't update if only one of the object entry is updated.
Because it needs to listen to each entry.

File: venia-integration-tests/src/fields/contactPage/index.js
Comment (pedrochiossi) [inline venia-integration-tests/src/fields/contactPage/index.js:1]: Why remove this selector?

**Review by pedrochiossi — CHANGES_REQUESTED** [review]
Nicely done @mikhaelbois, I left some comments and suggestions. I think we'll need a sign-off from core team as well.

File: packages/peregrine/lib/talons/CartPage/PriceAdjustments/GiftOptions/useGiftOptions.js
Comment (pedrochiossi) [reply packages/peregrine/lib/talons/CartPage/PriceAdjustments/GiftOptions/useGiftOptions.js:85]: I see, ok.

**Comment by mikhaelbois** [comment]
I have found a race condition issue with the checkboxes.
I will take a look on the December 27-28.

**Review by pedrochiossi — APPROVED** [review]

**Comment by jeremycharron** [comment]
> I have found a race condition issue with the checkboxes. I will take a look on the December 27-28.
> 
> Code Review and QA can still be done in the meantime.

Racing condition has been fixed.
Code review / QA can resume.

**Comment by fooman** [comment]
Would have loved to see an attempt to make PriceAdjustments extensible. Currently the approach seems to be that each order total becomes hardcoded.

**Comment by mikhaelbois** [comment]
@fooman 
Thanks for the suggestion! I'll let the team know so we can plan a refactor of this section.

**Review by michaelyu0123 — APPROVED** [review]
Just a small change needed, fix it and it LGTM

File: packages/venia-ui/i18n/en_US.json
Comment (mikhaelbois) [reply packages/venia-ui/i18n/en_US.json:222]: @michaelyu0123 
The `price` variable is optional so we don't have an unnecessary space in the text if no price is provided.
The space is added when there's a price.

**Comment by ericeoeur** [comment]
This has passed all QA testing. Will be merging momentarily. Thanks!

**Pull request merged**
